programs/vote/src/vote_state/mod.rs

```diff
@@ -1196,82 +1206,206 @@ mod tests {
+            let original_v4 = vote_state.as_ref_v4();
+            let converted_v4 = converted_vote_state.as_ref_v4();
+
+            // Inflation collector is set to the vote pubkey.
+            assert_eq!(converted_v4.inflation_rewards_collector, vote_pubkey);
+
+            // Remaining fields should be identical.
+            assert_eq!(original_v4.node_pubkey, converted_v4.node_pubkey);
+            assert_eq!(
+                original_v4.authorized_withdrawer,
+                converted_v4.authorized_withdrawer
+            );
+            assert_eq!(
+                original_v4.block_revenue_collector,
+                converted_v4.block_revenue_collector
+            );
+            assert_eq!(
+                original_v4.inflation_rewards_commission_bps,
+                converted_v4.inflation_rewards_commission_bps
+            );
+            assert_eq!(
+                original_v4.block_revenue_commission_bps,
+                converted_v4.block_revenue_commission_bps
+            );
+            assert_eq!(
+                original_v4.pending_delegator_rewards,
+                converted_v4.pending_delegator_rewards
+            );
+            assert_eq!(
+                original_v4.bls_pubkey_compressed,
+                converted_v4.bls_pubkey_compressed
+            );
+            assert_eq!(original_v4.votes, converted_v4.votes);
+            assert_eq!(original_v4.root_slot, converted_v4.root_slot);
+            assert_eq!(
+                original_v4.authorized_voters,
+                converted_v4.authorized_voters
+            );
+            assert_eq!(original_v4.epoch_credits, converted_v4.epoch_credits);
+            assert_eq!(original_v4.last_timestamp, converted_v4.last_timestamp);
+        };
 
         // Ensure that the vote state started out at 1_14_11
         let vote_state_version = borrowed_account.get_state::<VoteStateVersions>().unwrap();
         assert_matches!(vote_state_version, VoteStateVersions::V1_14_11(_));
 
         // Convert the vote state to current as would occur during vote instructions
-        let converted_vote_state = VoteStateV3::deserialize(borrowed_account.get_data()).unwrap();
+        let converted_vote_state =
+            VoteStateHandler::deserialize_and_convert(&borrowed_account, target_version).unwrap();
 
         // Check to make sure that the vote_state is unchanged
-        assert!(vote_state == converted_vote_state);
+        match target_version {
+            VoteStateTargetVersion::V3 => {
+                // V1_14_11 and V3 share the same fields, so the vote states
+                // should be identical.
+                assert!(vote_state == converted_vote_state);
+            }
+            VoteStateTargetVersion::V4 => {
+                // V4 has fields that V1_14_11 does not have, so the vote states
+                // will vary slightly.
+                check_converted_vote_state_v4_fields(&converted_vote_state);
```

Comment:
Why not have the `vote_state_new_for_test` helper take a `vote_pubkey` param which can be used to initialize the inflation collector address? That way we can just do `assert!(vote_state == converted_vote_state);` no matter the target version.